src/inngest/functions/backfillUserActionRefer/index.ts

```diff
@@ -0,0 +1,180 @@
+import { DataCreationMethod, UserActionType } from '@prisma/client'
+import { groupBy } from 'lodash-es'
+
+import { inngest } from '@/inngest/inngest'
+import { onScriptFailure } from '@/inngest/onScriptFailure'
+import { prismaClient } from '@/utils/server/prismaClient'
+import { batchAsyncAndLog } from '@/utils/shared/batchAsyncAndLog'
+import { USER_ACTION_TO_CAMPAIGN_NAME_DEFAULT_MAP } from '@/utils/shared/userActionCampaigns'
+
+export const BACKFILL_USER_ACTION_REFER_INNGEST_EVENT_NAME = 'script/backfill-user-action-refer'
+export const BACKFILL_USER_ACTION_REFER_INNGEST_FUNCTION_ID = 'script.backfill-user-action-refer'
+
+export interface BackfillUserActionReferInngestSchema {
+  name: typeof BACKFILL_USER_ACTION_REFER_INNGEST_EVENT_NAME
+  data: {
+    persist: boolean
+  }
+}
+
+const selectFields = {
+  id: true,
+  userId: true,
+  actionType: true,
+  campaignName: true,
+  dataCreationMethod: true,
+  userActionRefer: {
+    select: {
+      referralsCount: true,
+    },
+  },
+}
+
+/**
+ * This function is used to backfill the user action refer table.
+ * It will find all users who were referred (have utm_source=swc and utm_medium=referral)
+ * and then find all referrers by their referral IDs.
+ * It will then create or update the referrer's user action refer record.
+ */
+export const backfillUserActionRefer = inngest.createFunction(
+  {
+    id: BACKFILL_USER_ACTION_REFER_INNGEST_FUNCTION_ID,
+    retries: 0,
+    onFailure: onScriptFailure,
+  },
+  { event: BACKFILL_USER_ACTION_REFER_INNGEST_EVENT_NAME },
+  async ({ event, step, logger }) => {
+    const { persist = false } = event.data
+
+    logger.info(`Started with persist=${String(persist)}`)
+
+    // Find all users who were referred (have utm_source=swc and utm_medium=referral)
+    const referredUsers = await step.run('find-referred-users', async () => {
+      const users = await prismaClient.user.findMany({
+        where: {
+          acquisitionSource: 'swc',
+          acquisitionMedium: 'referral',
+          NOT: {
+            acquisitionCampaign: '',
+          },
+        },
+        select: {
+          acquisitionCampaign: true,
+        },
+      })
+      logger.info(`Found ${users.length} referred users`)
+      return users
```

Comment:
Not a problem, bu since you're only using `referredUsers` on the groupBy, you could return it already grouped

```suggestion
      return groupBy(users, 'acquisitionCampaign')
```